_posts/2021-12-3-arrow-duckdb.md

```diff
@@ -0,0 +1,383 @@
+
+### Filter Pushdown
+
+For our filter pushdown we repeat the same aggregation used in the previous section, but add filters on 4 more columns.
+
+``` python
+# DuckDB
+lineitem = pq.read_table('lineitemsf1.snappy.parquet')
+
+# Get database connection
+con = duckdb.connect()
+
+# Transforms Query Result from DuckDB to Arrow Table
+con.execute("""SELECT sum(l_extendedprice * l_discount) AS revenue
+        FROM
+            lineitem
+        WHERE
+            l_shipdate >= CAST('1994-01-01' AS date)
+            AND l_shipdate < CAST('1995-01-01' AS date)
+            AND l_discount BETWEEN 0.05
+            AND 0.07
+            AND l_quantity < 24; """).fetch_arrow_table()
+
+```
+
+``` python
+# Pandas
+arrow_table = pq.read_table('lineitemsf1.snappy.parquet')
+
+df = arrow_table.to_pandas()
+filtered_df = lineitem[
+        (lineitem.l_shipdate >= "1994-01-01") &
+        (lineitem.l_shipdate < "1995-01-01") &
+        (lineitem.l_discount >= 0.05) &
+        (lineitem.l_discount <= 0.07) &
+        (lineitem.l_quantity < 24)]
+
+res =  pd.DataFrame({'sum': [(filtered_df.l_extendedprice * filtered_df.l_discount).sum()]})
+new_table = pa.Table.from_pandas(res)
+```
+
+|    Name     | Time (s) |
+|-------------|----------|
+| DuckDB  | 0.04    |
+| Pandas      | 2.29    |
+
+The difference now between DuckDB and Pandas is more drastic, being two orders of magnitude faster than Pandas. Again, since both the filter and projection are pushed down to Arrow, DuckDB reads less data than Pandas, which can't automatically perform this optimization.
+
+### Streaming
+
+As demonstrated before, DuckDB is capable of consuming and producing Arrow data in a streaming fashion. In this section we run a simple benchmark, to showcase the benefits in speed and memory usage when comparing it to full materialization and Pandas. This example uses the full NYC taxi dataset which you can download
+
+
+``` python
+# DuckDB
+# Open dataset using year,month folder partition
+nyc = ds.dataset('nyc-taxi/', partitioning=["year", "month"])
+
+# Get database connection
+con = duckdb.connect()
+
+# Run query that selects part of the data
+query = con.execute("SELECT total_amount, passenger_count,year FROM nyc where total_amount > 100 and year > 2014")
+
+# Create Record Batch Reader from Query Result.
+# "fetch_record_batch()" also accepts an extra parameter related to the desired produced chunk size.
+record_batch_reader = query.fetch_record_batch()
+
+# Retrieve all batch chunks
+chunk = record_batch_reader.read_next_batch()
+while (len(chunk) > 0):
+    chunk = record_batch_reader.read_next_batch()
+```
+
+``` python
+# Pandas
+# We must exclude one of the columns of the NYC dataset due to an unimplemented cast in Arrow.
```

Comment:
(side note: I think you can fix this by telling Arrow to scan more than one file to determine the schema)

Reply:
Yeah, specifying unify_schemas works, though scans more. We also have plans to (finally) clean this inconsistency up at some point!